### ISSUE TITLE
Add configuration for no-did-update-set-state to allow usage inside condition

### ISSUE THREAD

**Issue by dcocchia** [yannickcr/eslint-plugin-react#2021]
As @gaearon has mentioned [here](https://github.com/yannickcr/eslint-plugin-react/issues/1059#issuecomment-276828605) and other places, using `setState` inside `componenDidUpdate` is acceptable, so long as it's within a conditional check. 

Does it seem reasonable to allow a configuration option for this rule to allow `setState` usage within `componenDidUpdate`, so long as it's within a conditional check of some kind?

For example, this would pass here:
```js
componentDidUpdate() {
  if (!this.state.isEnabled) {
    this.setTate({ isEnabled: true })
  }
}
```

but this would still fail:
```js
componentDidUpdate() {
  this.setTate({ isEnabled: true })
}
```

An issue here would be that it's still not entirely guaranteed that more complex use cases wouldn't still end up in an infinite loop.

**Comment by himynameisdave** [comment]
This would be awesome. I think it's fine if it's in a conditional or if you move the `setState` call off to another method.

**Comment by devourment77** [comment]
In the react docs for [componentDidUpdate](https://reactjs.org/docs/react-component.html#componentdidupdate), it says `You may call setState() immediately in componentDidUpdate() but note that it must be wrapped in a condition` and the example is an if statement.

**Comment by ljharb** [comment]
The condition would have to be one that prevents an infinite loop, which isn’t something that’s be statically knowable.

**Comment by himynameisdave** [comment]
@ljharb true, but I still want to use this rule. Maybe I'll just change it to a warning in our codebase, but overall I prefer rules to be strict so that they cannot be committed at all.

**Comment by gaurav5430** [comment]
any updates on whether this is going to be an option or not ?
